src/field.jl

```diff
@@ -121,55 +121,137 @@ fromtype(::Type{N}) where {N<:Number} = Numeric(N)
 ### Unsigned integers
 
 function tryparsenext(::Numeric{T}, str, i, len) where {T<:Signed}
+    R = Nullable{T}
     @chk2 sign, i = tryparsenext_sign(str, i, len)
     @chk2 x, i = tryparsenext_base10(T, str, i, len)
 
     @label done
-    return Some(convert(T, sign*x)), i
+    return R(convert(T, sign*x)), i
 
     @label error
-    return nothing, i
+    return R(), i
 end
 
 @inline function tryparsenext(::Numeric{T}, str, i, len) where {T<:Unsigned}
     tryparsenext_base10(T,str, i, len)
 end
 
+@inline _is_e(str, i) = str[i]=='e' || str[i]=='E'
+
+@inline _is_negative(str, i) = str[i]=='-'
+
+@inline _is_positive(str, i) = str[i]=='+'
+
+@inline function convert_to_double(f1::Int64, exp::Int)
+    f = Float64(f1)
+    r = f1 - Int64(f) # get the remainder
+    x = Double64(f) + Double64(r)
+  
+    maxexp = 308
+    minexp = -256
+  
+    if exp >= 0
+        x *= Double64(10.0)^(exp)
+    else
+        if exp < minexp # not sure why this is a good choice, but it seems to be!
+            x /= Double64(10.0)^(-minexp)
+            x /= Double64(10.0)^(-exp + minexp)
+        else
+            x /= Double64(10.0)^(-exp)
+        end
+    end
+    return Float64(x)
+end
```

Comment:
@simonbyrne would be great if you could indicate whether you will be able to review this function in particular in the near future. Totally understood if not, but in that case I think I would just merge the PR, given that it passes all tests, and passes more parsing tests than the current ``master`` version. Even if there is something wrong with this code, I think it would still be better than the status quo. BUT, a review from you would be really great, of course!

And if you have too much time (haha), you could also take a quick look at the function that calls this one here, I think at that point you would have reviewed all that there is to this float parsing.

Reply:
I had a bit of  a look over: although it looks like it will help a bit, it won't be able to completely solve the problem ([there are some cases which are beyond double-double](https://github.com/JuliaData/Parsers.jl/issues/5#issuecomment-422908055)).

It will also likely be much slower: ideally you only want to utilise excess precision only when you need to. Also it probably makes sense to handle the powers of 10 via a lookup table (possible two separate ones: one where 10.0^k is exact, and another in higher precision).

Reply:
> although it looks like it will help a bit

Well, the issue previously was that if you just generated a bunch of random numbers, wrote them out to a string (with the default number of digits in julia write methods) and then read them back in, about 10% of those were read incorrectly. In my mind that was a large fraction... This PR seems to mostly solve that, at least for the random number streams I tested, this goes to 0 errors.

> it won't be able to completely solve the problem

Thanks, that is good to know. So my take is that we now go from reading about 10% incorrectly to a number of tricky corner cases. We should try to fix them as well, but for now I'm happy enough with the improvements here :)

I guess the main question here is whether there is an easy way to detect situations where even the double double code path is not good enough and we need to switch to something else for those corner cases...

> It will also likely be much slower: ideally you only want to utilise excess precision only when you need to.

So the double double code path is only taken if needed, otherwise a faster code path runs (see [here](https://github.com/JuliaComputing/TextParse.jl/pull/74/files#diff-dbdafaad817e6086a65af08fd65ccfa5R172)). But even the double double case shows good performance: if you look at the aggregate performance for the ``uniform_data_float64.csv`` file in the benchmark I posted, performance in this PR (using the double double code path) is actually *better* than in TextParse.jl on julia 0.6 (which didn't use any extended precision at all). Relative to all the other packages, only R fread beats it (but that beats us in almost all cases).

> Also it probably makes sense to handle the powers of 10 via a lookup table (possible two separate ones: one where 10.0^k is exact, and another in higher precision).

I played around with that at some point and didn't see any improvements. I didn't do it super carefully, might well have been that I got something wrong and it would have helped if done more carefully.

Reply:
> So the double double code path is only taken if needed, otherwise a faster code path runs 

Ah, I missed how that worked. In that case, it is certainly a net improvement.